### PR TITLE
PROTOCOLS-117 MailboxName LIST filtering should fully be a PROTOCOL layer concern

### PR DESCRIPTION
**Pull request by chibenwa**
This PR is a simple constat:

IMAP expects  filtering on the **name exposed to the client** (mailbox name), not on the **internal name representation** (mailbox path).

This is OK as long as a direct mapping is maintained between mailbox name and path (what is doable as long as you keep a single mapping between MailboxPath and mailbox name). But when you starts wanting doing things like:

```
# Current
a0 LIST "" "#private.bob.*"
a0 LIST "" "Other users.bob.**"
```

Things breaks ;-) That's why I concluded MailboxName matching is a protocol level concern. They, BTW, were no other uses.

### CODE REVIEW

**Comment by mbaechler** [comment]
could we get rid of BONUS tag in commit titles ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java [hunk 1]
```diff
@@ -644,11 +643,7 @@ private void doRenameMailbox(MailboxPath from, MailboxPath to, MailboxSession se
 
     @VisibleForTesting
     public static MailboxPath getPathLike(MailboxQuery mailboxQuery, MailboxSession mailboxSession) {
-        MailboxNameExpression nameExpression = mailboxQuery.getMailboxNameExpression();
-        String combinedName = nameExpression.getCombinedName()
-            .replace(nameExpression.getFreeWildcard(), SQL_WILDCARD_CHAR)
-            .replace(nameExpression.getLocalWildcard(), SQL_WILDCARD_CHAR)
-            + SQL_WILDCARD_CHAR;
+        String combinedName = String.valueOf(SQL_WILDCARD_CHAR);
```
Comment (mbaechler) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMailboxManager.java]: why it's still called combinedName ?

File: protocols/imap/src/main/java/org/apache/james/imap/processor/ListProcessor.java [hunk 2]
```diff
@@ -172,15 +175,19 @@ public MailboxId getId() {
                     basePath = PathConverter.forSession(session).buildFullPath(finalReferencename);
                 }
 
-                results = getMailboxManager().search(
```
Comment (mbaechler) [inline protocols/imap/src/main/java/org/apache/james/imap/processor/ListProcessor.java]: is this part well covered by test ?

File: protocols/imap/src/main/java/org/apache/james/imap/processor/ListProcessor.java [hunk 3]
```diff
@@ -172,15 +175,19 @@ public MailboxId getId() {
...
+                PrefixedRegex prefixedRegex = new PrefixedRegex(
+                    basePath.getName(),
+                    CharsetUtil.decodeModifiedUTF7(mailboxName),
+                    mailboxSession.getPathDelimiter());
+                results = getMailboxManager()
```
Comment (mbaechler) [inline protocols/imap/src/main/java/org/apache/james/imap/processor/ListProcessor.java]: are we loading every single user mailboxes the user has access to then run a java filter on the collection ?

File: protocols/imap/src/main/java/org/apache/james/imap/processor/ListProcessor.java (hunk 2 above)
Comment (chibenwa) [reply protocols/imap/src/main/java/org/apache/james/imap/processor/ListProcessor.java]: Yes, in MPT

File: protocols/imap/src/main/java/org/apache/james/imap/processor/ListProcessor.java (hunk 3 above)
Comment (chibenwa) [reply protocols/imap/src/main/java/org/apache/james/imap/processor/ListProcessor.java]: Yes, but it's loaded anyway. And it is not that bad.

BTW we can speak of why this is needed ;-) I would send you back to PR description.

Comment (rouazana) [reply protocols/imap/src/main/java/org/apache/james/imap/processor/ListProcessor.java]: I understand that Cassandra is already doing it, but what about the other backends? JPA for example?

Comment (chibenwa) [reply protocols/imap/src/main/java/org/apache/james/imap/processor/ListProcessor.java]: Yes, JPA will be less optimized.

That being said we don't know how to handle push back reference to make this both **correct** and **optimized** (push back reference if I remember @mbaechler  last daily speach). In that context, we prefered doing the **correct** choice, even at a performance penalty.

Comment (mbaechler) [reply protocols/imap/src/main/java/org/apache/james/imap/processor/ListProcessor.java]: >  In that context, we prefered doing the correct choice, even at a performance penalty.

"We" at least doesn't include "me". I agreed having same performance property for cassandra. I didn't agree to make JPA worse.

**Comment by mbaechler** [comment]
Ok, I think we should go back to basics here. If I understand correctly, the problem is that a MailboxPath can have several representation in IMAP world (we will name it ImapMailboxPath for now, whether it's a good name or not).
So, the problem is AFAIK that for a mailbox A, we can have two ImapMailboxPath : B and C.
When the user asks to filter on a prefix of B for example, and C doesn't have the same prefix, it shouldn't filter C from the output.
But AFAIK, _for a given user_ there is a bijection between A and B or A and C, right ?
Then what prevents from first building the MailboxPath for the user then do the filtering in the backend ?
Please correct me if I'm wrong.

**Comment by chibenwa** [comment]
Ok.

Having thought about it:

If you want your storage to handle it, **you need context aware storage** that would play *name matching* against the context of the user.

Needless to say, we don't have context aware storage.

That means given:

```
#private:mbaechler:INBOX.folder
```

It needs to be stored as:

```
INBOX.folder
```

In the context of matthieu

And

```
Other users.mbaechler.INBOX.folder
```

In my context.

Of course it is doable. Of course it is **IMAP specific**. Now **how to expose IMAP specifc APIs** on a generic store interface?

My bet, if we want to go that way:

 - A MailboxListener should handle **context denormalization** for each user. That means storing the mailbox name as seen by each user.
 - Introduce a search API based on this context.

But **that is specific to IMAP**, which means such components **needs to be defined in IMAP project**, maybe with **data-imap** the same way than **what we already do in JMAP**. 

My opinion: take IMAP specific stuff out of Store. That just makes things harder to have IMAP concerns leaking everywhere.

**Comment by aduprat** [comment]
Build triggered. sha1 is merged.

**Comment by aduprat** [comment]
Build started sha1 is merged.

**Comment by aduprat** [comment]
Build finished. No test results found.

**Comment by aduprat** [comment]
Build triggered. sha1 is merged.

**Comment by aduprat** [comment]
Build started sha1 is merged.

**Comment by aduprat** [comment]
Build finished. No test results found.

**Comment by aduprat** [comment]
Build triggered. sha1 is merged.

**Comment by aduprat** [comment]
Build started sha1 is merged.

**Comment by aduprat** [comment]
Build finished. No test results found.

**Comment by openpaas-james** [comment]
Build triggered. sha1 is original commit.

**Comment by openpaas-james** [comment]
Build started sha1 is original commit.

**Comment by openpaas-james** [comment]
Build finished. No test results found.

**Comment by chibenwa** [comment]
Delegating support for IMAP is not a priority and won't be delivered any time soon

I'm closing this long outdated work.

**Pull request closed without merging**